src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,936 @@
+
+	char *newColocateWith = con->colocateWith;
+	if (con->colocateWith == NULL)
+	{
+		if (con->originalDistributionKey->vartype == newDistributionKey->vartype &&
+			con->shardCountIsNull)
+		{
+			newColocateWith =
+				quote_qualified_identifier(con->schemaName, con->relationName);
+		}
+		else
+		{
+			newColocateWith = "default";
+		}
+	}
+	char partitionMethod = PartitionMethod(con->relationId);
+	CreateDistributedTable(con->newRelationId, newDistributionKey, partitionMethod,
+						   con->shardCount, newColocateWith, false);
+}
+
+
+void
+CreateCitusTableLike(TableConversion *con)
+{
+	if (IsCitusTableType(con->relationId, DISTRIBUTED_TABLE))
+	{
+		CreateDistributedTableLike(con);
+	}
+	else if (IsCitusTableType(con->relationId, REFERENCE_TABLE))
+	{
+		CreateDistributedTable(con->newRelationId, NULL, DISTRIBUTE_BY_NONE, 0,
+							   NULL, false);
+	}
+	else if (IsCitusTableType(con->relationId, CITUS_LOCAL_TABLE))
+	{
+		CreateCitusLocalTable(con->newRelationId);
+	}
+}
+
+
+/*
+ * GetViewCreationCommandsOfTable takes a table oid generates the CREATE VIEW
+ * commands for views that depend to the given table. This includes the views
+ * that recursively depend on the table too.
+ */
+List *
+GetViewCreationCommandsOfTable(Oid relationId)
+{
+	List *views = GetDependingViews(relationId);
+	List *commands = NIL;
+
+	Oid viewOid = InvalidOid;
+	foreach_oid(viewOid, views)
+	{
+		Datum viewDefinitionDatum = DirectFunctionCall1(pg_get_viewdef,
+														ObjectIdGetDatum(viewOid));
+		char *viewDefinition = TextDatumGetCString(viewDefinitionDatum);
+		StringInfo query = makeStringInfo();
+		char *viewName = get_rel_name(viewOid);
+		char *schemaName = get_namespace_name(get_rel_namespace(viewOid));
+		char *qualifiedViewName = quote_qualified_identifier(schemaName, viewName);
+		appendStringInfo(query,
+						 "CREATE VIEW %s AS %s",
+						 qualifiedViewName,
+						 viewDefinition);
+		commands = lappend(commands, makeTableDDLCommandString(query->data));
+	}
+	return commands;
+}
+
+
+/*
+ * ReplaceTable replaces the source table with the target table.
+ * It moves all the rows of the source table to target table with INSERT SELECT.
+ * Changes the dependencies of the sequences owned by source table to target table.
+ * Then drops the source table and renames the target table to source tables name.
+ *
+ * Source and target tables need to be in the same schema and have the same columns.
+ */
+void
+ReplaceTable(Oid sourceId, Oid targetId, List *justBeforeDropCommands)
+{
+	char *sourceName = get_rel_name(sourceId);
+	char *targetName = get_rel_name(targetId);
+	Oid schemaId = get_rel_namespace(sourceId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	StringInfo query = makeStringInfo();
+
+	ereport(NOTICE, (errmsg("Moving the data of %s",
+							quote_qualified_identifier(schemaName, sourceName))));
+
+	appendStringInfo(query, "INSERT INTO %s SELECT * FROM %s",
```

Comment:
In case of partitioned tables, we seem to do this both for the parent and for the children. The reason it does not result in duplicate data, is that all the children have been detached at this point. It might be a little safer to skip this for the parent.

Reply:
(I also kind of wonder whether it would be better to recreate the partition hierarchy and then do the INSERT..SELECT and drop the whole hierarchy at once, which also avoids blocking reads early on, but that might be more complicated, so only nice-to-have

Reply:
I'm not super familiar with partitioned tables. I thought either there is no data in the parent, so the `INSERT SELECT` will not do anything, or there can be some data in the parent, so the `INSERT SELECT` is necessary. Judging from your comment I assume the second cannot be the case.

I will think about the creating and dropping hierarchy at once but I agree it is a bit complicated. So maybe could be a separate PR. 😃

Reply:
There cannot be data in the parent, but when selecting from the parent it will include all the data from the parent.